File: v2/validation/filters.ts
/**
 * Filter Validation Module
 *
 * Config-driven validation for filter types.
 * Checks that only supported filter types are used.
 */

import { ValidationConfig } from '../config/types';

// =============================================================================
// TYPES
// =============================================================================

export interface FilterValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
  usedFilters: {
    string: string[];
    numeric: string[];
    null: string[];
  };
}

// =============================================================================
// FILTER TYPE EXTRACTORS
// =============================================================================

/**
 * Extract all string filter types used in a filter (recursively)
 */
export function extractStringFilterTypes(filter: unknown): Set<string> {
  const types = new Set<string>();

  if (!filter || typeof filter !== 'object') return types;

  const f = filter as Record<string, unknown>;

  // Column filter with string type
  if (f.columnFilterRule) {
    const rule = f.columnFilterRule as Record<string, unknown>;
    const columnFilter = rule.filter as Record<string, unknown>;
    const strFilter = columnFilter?.stringColumnFilter as Record<string, unknown>;
    if (strFilter?.type) {
      types.add(strFilter.type as string);
    }
  }

  // Recurse into compound filters
  if (f.orFilterRule) {
    const rule = f.orFilterRule as Record<string, unknown>;
    const filters = rule.filters as unknown[];
    if (Array.isArray(filters)) {
      for (const subFilter of filters) {
        extractStringFilterTypes(subFilter).forEach(t => types.add(t));
      }
    }
  }

  if (f.andFilterRule) {
    const rule = f.andFilterRule as Record<string, unknown>;
    const filters = rule.filters as unknown[];
    if (Array.isArray(filters)) {
      for (const subFilter of filters) {
        extractStringFilterTypes(subFilter).forEach(t => types.add(t));
      }
    }
  }

  if (f.notFilterRule) {
    const rule = f.notFilterRule as Record<string, unknown>;
    extractStringFilterTypes(rule.filter).forEach(t => types.add(t));
  }

  return types;
}

/**
 * Extract all numeric filter types used in a filter (recursively)
 */
export function extractNumericFilterTypes(filter: unknown): Set<string> {
  const types = new Set<string>();

  if (!filter || typeof filter !== 'object') return types;

  const f = filter as Record<string, unknown>;

  // Column filter with numeric type
  if (f.columnFilterRule) {
    const rule = f.columnFilterRule as Record<string, unknown>;
    const columnFilter = rule.filter as Record<string, unknown>;
    const numFilter = columnFilter?.numericColumnFilter as Record<string, unknown>;
    if (numFilter?.type) {
      types.add(numFilter.type as string);
    }
  }

  // Recurse into compound filters
  if (f.orFilterRule) {
    const rule = f.orFilterRule as Record<string, unknown>;
    const filters = rule.filters as unknown[];
    if (Array.isArray(filters)) {
      for (const subFilter of filters) {
        extractNumericFilterTypes(subFilter).forEach(t => types.add(t));
      }
    }
  }

  if (f.andFilterRule) {
    const rule = f.andFilterRule as Record<string, unknown>;
    const filters = rule.filters as unknown[];
    if (Array.isArray(filters)) {
      for (const subFilter of filters) {
        extractNumericFilterTypes(subFilter).forEach(t => types.add(t));
      }
    }
  }

  if (f.notFilterRule) {
    const rule = f.notFilterRule as Record<string, unknown>;
    extractNumericFilterTypes(rule.filter).forEach(t => types.add(t));
  }

  return types;
}

/**
 * Extract all null filter types used in a filter (recursively)
 */
export function extractNullFilterTypes(filter: unknown): Set<string> {
  const types = new Set<string>();

  if (!filter || typeof filter !== 'object') return types;

  const f = filter as Record<string, unknown>;

  // Column filter with null type
  if (f.columnFilterRule) {
    const rule = f.columnFilterRule as Record<string, unknown>;
    const columnFilter = rule.filter as Record<string, unknown>;
    const nullFilter = columnFilter?.nullColumnFilter as Record<string, unknown>;
    if (nullFilter?.type) {
      types.add(nullFilter.type as string);
    }
  }

  // Recurse into compound filters
  if (f.orFilterRule) {
    const rule = f.orFilterRule as Record<string, unknown>;
    const filters = rule.filters as unknown[];
    if (Array.isArray(filters)) {
      for (const subFilter of filters) {
        extractNullFilterTypes(subFilter).forEach(t => types.add(t));
      }
    }
  }

  if (f.andFilterRule) {
    const rule = f.andFilterRule as Record<string, unknown>;
    const filters = rule.filters as unknown[];
    if (Array.isArray(filters)) {
      for (const subFilter of filters) {
        extractNullFilterTypes(subFilter).forEach(t => types.add(t));
      }
    }
  }

  if (f.notFilterRule) {
    const rule = f.notFilterRule as Record<string, unknown>;
    extractNullFilterTypes(rule.filter).forEach(t => types.add(t));
  }

  return types;
}

// =============================================================================
// VALIDATION
// =============================================================================

/**
 * Validate filter types against config-driven rules
 */
export function validateFilterTypes(
  logic: unknown,
  validationConfig: ValidationConfig
): FilterValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  // Extract filter from rule logic
  const l = logic as Record<string, unknown>;
  const strategy = l?.strategy as Record<string, unknown>;
  const filterNode = strategy?.filterNode as Record<string, unknown>;
  const filter = filterNode?.filter;

  // Extract all filter types used
  const stringTypes = Array.from(extractStringFilterTypes(filter));
  const numericTypes = Array.from(extractNumericFilterTypes(filter));
  const nullTypes = Array.from(extractNullFilterTypes(filter));

  // Validate string filters
  for (const filterType of stringTypes) {
    if (validationConfig.unsupportedStringFilters?.includes(filterType)) {
      errors.push(
        `String filter type "${filterType}" is not supported. ` +
        `Use one of: ${validationConfig.supportedStringFilters.join(', ')}`
      );
    } else if (!validationConfig.supportedStringFilters.includes(filterType)) {
      warnings.push(
        `String filter type "${filterType}" is not in the list of tested filters. ` +
        `It may not render correctly in the UI.`
      );
    }
  }

  // Validate numeric filters
  if (validationConfig.supportedNumericFilters?.length) {
    for (const filterType of numericTypes) {
      if (!validationConfig.supportedNumericFilters.includes(filterType)) {
        warnings.push(
          `Numeric filter type "${filterType}" is not in the list of tested filters.`
        );
      }
    }
  }

  // Validate null filters
  if (validationConfig.supportedNullFilters?.length) {
    for (const filterType of nullTypes) {
      if (!validationConfig.supportedNullFilters.includes(filterType)) {
        warnings.push(
          `Null filter type "${filterType}" is not in the list of tested filters.`
        );
      }
    }
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings,
    usedFilters: {
      string: stringTypes,
      numeric: numericTypes,
      null: nullTypes
    }
  };
}

/**
 * Check common filter type mistakes and suggest corrections
 */
export function suggestCorrections(filterType: string): string | null {
  const corrections: Record<string, string> = {
    'REGEX': 'MATCHES',
    'WILDCARD': 'EQUALS_WITH_WILDCARDS',
    'IS_NULL': 'NULL',
    'IS_NOT_NULL': 'NOT_NULL',
    'GREATER_THAN_OR_EQUALS': 'GREATER_THAN_OR_EQUAL',
    'LESS_THAN_OR_EQUALS': 'LESS_THAN_OR_EQUAL',
    'GTE': 'GREATER_THAN_OR_EQUAL',
    'LTE': 'LESS_THAN_OR_EQUAL',
    'GT': 'GREATER_THAN',
    'LT': 'LESS_THAN',
    'EQ': 'EQUALS',
    'NEQ': 'Use NOT(EQUALS) instead',
    'NOT_EQUALS': 'Use NOT(EQUALS) instead'
  };

  return corrections[filterType] || null;
}

/**
 * Get ignoreWhitespace requirement for string filter types
 */
export function requiresIgnoreWhitespace(filterType: string): boolean {
  // These types require ignoreWhitespace: true to work correctly
  return ['CONTAINS', 'MATCHES', 'EQUALS_WITH_WILDCARDS'].includes(filterType);
}

/**
 * Validate ignoreWhitespace settings in string filters
 */
export function validateIgnoreWhitespace(filter: unknown): {
  valid: boolean;
  errors: string[];
} {
  const errors: string[] = [];

  validateIgnoreWhitespaceRecursive(filter, errors);

  return { valid: errors.length === 0, errors };
}

function validateIgnoreWhitespaceRecursive(filter: unknown, errors: string[]): void {
  if (!filter || typeof filter !== 'object') return;

  const f = filter as Record<string, unknown>;

  // Check string column filter
  if (f.columnFilterRule) {
    const rule = f.columnFilterRule as Record<string, unknown>;
    const columnFilter = rule.filter as Record<string, unknown>;
    const strFilter = columnFilter?.stringColumnFilter as Record<string, unknown>;

    if (strFilter) {
      const filterType = strFilter.type as string;
      const ignoreWhitespace = strFilter.ignoreWhitespace as boolean;

      if (requiresIgnoreWhitespace(filterType) && ignoreWhitespace !== true) {
        errors.push(
          `String filter type "${filterType}" requires ignoreWhitespace: true, ` +
          `but got: ${ignoreWhitespace}`
        );
      }
    }
  }

  // Recurse into compound filters
  if (f.orFilterRule) {
    const rule = f.orFilterRule as Record<string, unknown>;
    const filters = rule.filters as unknown[];
    if (Array.isArray(filters)) {
      for (const subFilter of filters) {
        validateIgnoreWhitespaceRecursive(subFilter, errors);
      }
    }
  }

  if (f.andFilterRule) {
    const rule = f.andFilterRule as Record<string, unknown>;
    const filters = rule.filters as unknown[];
    if (Array.isArray(filters)) {
      for (const subFilter of filters) {
        validateIgnoreWhitespaceRecursive(subFilter, errors);
      }
    }
  }

  if (f.notFilterRule) {
    const rule = f.notFilterRule as Record<string, unknown>;
    validateIgnoreWhitespaceRecursive(rule.filter, errors);
  }
}

/**
 * Get a summary of all filter types used
 */
export function getFilterSummary(logic: unknown): {
  stringFilters: string[];
  numericFilters: string[];
  nullFilters: string[];
  hasCompoundFilters: boolean;
} {
  const l = logic as Record<string, unknown>;
  const strategy = l?.strategy as Record<string, unknown>;
  const filterNode = strategy?.filterNode as Record<string, unknown>;
  const filter = filterNode?.filter as Record<string, unknown>;

  return {
    stringFilters: Array.from(extractStringFilterTypes(filter)),
    numericFilters: Array.from(extractNumericFilterTypes(filter)),
    nullFilters: Array.from(extractNullFilterTypes(filter)),
    hasCompoundFilters: !!(filter?.orFilterRule || filter?.andFilterRule || filter?.notFilterRule)
  };
}

File: src/foundry_rules/validation/filters.py
"""
Filter Validation Module

Config-driven validation for filter types.
"""

from typing import Any
from pydantic import BaseModel

from ..config.types import ValidationConfig


class FilterValidationResult(BaseModel):
    """Result of filter validation."""

    valid: bool
    errors: list[str] = []
    warnings: list[str] = []
    used_filters: dict[str, list[str]] = {}


def extract_string_filter_types(filter_obj: Any) -> set[str]:
    """Extract all string filter types used in a filter (recursively)."""
    types: set[str] = set()

    if not filter_obj or not isinstance(filter_obj, dict):
        return types

    # Column filter with string type
    if "columnFilterRule" in filter_obj:
        rule = filter_obj["columnFilterRule"]
        if isinstance(rule, dict):
            column_filter = rule.get("filter")
            if column_filter and isinstance(column_filter, dict):
                str_filter = column_filter.get("stringColumnFilter")
                if str_filter and isinstance(str_filter, dict):
                    filter_type = str_filter.get("type")
                    if filter_type:
                        types.add(filter_type)

    # Recurse into compound filters
    if "orFilterRule" in filter_obj:
        rule = filter_obj["orFilterRule"]
        if isinstance(rule, dict):
            filters = rule.get("filters", [])
            if isinstance(filters, list):
                for sub_filter in filters:
                    types.update(extract_string_filter_types(sub_filter))

    if "andFilterRule" in filter_obj:
        rule = filter_obj["andFilterRule"]
        if isinstance(rule, dict):
            filters = rule.get("filters", [])
            if isinstance(filters, list):
                for sub_filter in filters:
                    types.update(extract_string_filter_types(sub_filter))

    if "notFilterRule" in filter_obj:
        rule = filter_obj["notFilterRule"]
        if isinstance(rule, dict):
            types.update(extract_string_filter_types(rule.get("filter")))

    return types


def extract_numeric_filter_types(filter_obj: Any) -> set[str]:
    """Extract all numeric filter types used in a filter (recursively)."""
    types: set[str] = set()

    if not filter_obj or not isinstance(filter_obj, dict):
        return types

    # Column filter with numeric type
    if "columnFilterRule" in filter_obj:
        rule = filter_obj["columnFilterRule"]
        if isinstance(rule, dict):
            column_filter = rule.get("filter")
            if column_filter and isinstance(column_filter, dict):
                num_filter = column_filter.get("numericColumnFilter")
                if num_filter and isinstance(num_filter, dict):
                    filter_type = num_filter.get("type")
                    if filter_type:
                        types.add(filter_type)

    # Recurse into compound filters
    if "orFilterRule" in filter_obj:
        rule = filter_obj["orFilterRule"]
        if isinstance(rule, dict):
            filters = rule.get("filters", [])
            if isinstance(filters, list):
                for sub_filter in filters:
                    types.update(extract_numeric_filter_types(sub_filter))

    if "andFilterRule" in filter_obj:
        rule = filter_obj["andFilterRule"]
        if isinstance(rule, dict):
            filters = rule.get("filters", [])
            if isinstance(filters, list):
                for sub_filter in filters:
                    types.update(extract_numeric_filter_types(sub_filter))

    if "notFilterRule" in filter_obj:
        rule = filter_obj["notFilterRule"]
        if isinstance(rule, dict):
            types.update(extract_numeric_filter_types(rule.get("filter")))

    return types


def extract_null_filter_types(filter_obj: Any) -> set[str]:
    """Extract all null filter types used in a filter (recursively)."""
    types: set[str] = set()

    if not filter_obj or not isinstance(filter_obj, dict):
        return types

    # Column filter with null type
    if "columnFilterRule" in filter_obj:
        rule = filter_obj["columnFilterRule"]
        if isinstance(rule, dict):
            column_filter = rule.get("filter")
            if column_filter and isinstance(column_filter, dict):
                null_filter = column_filter.get("nullColumnFilter")
                if null_filter and isinstance(null_filter, dict):
                    filter_type = null_filter.get("type")
                    if filter_type:
                        types.add(filter_type)

    # Recurse into compound filters
    if "orFilterRule" in filter_obj:
        rule = filter_obj["orFilterRule"]
        if isinstance(rule, dict):
            filters = rule.get("filters", [])
            if isinstance(filters, list):
                for sub_filter in filters:
                    types.update(extract_null_filter_types(sub_filter))

    if "andFilterRule" in filter_obj:
        rule = filter_obj["andFilterRule"]
        if isinstance(rule, dict):
            filters = rule.get("filters", [])
            if isinstance(filters, list):
                for sub_filter in filters:
                    types.update(extract_null_filter_types(sub_filter))

    if "notFilterRule" in filter_obj:
        rule = filter_obj["notFilterRule"]
        if isinstance(rule, dict):
            types.update(extract_null_filter_types(rule.get("filter")))

    return types


def validate_filter_types(
    logic: Any,
    validation_config: ValidationConfig,
) -> FilterValidationResult:
    """
    Validate filter types against config-driven rules.

    Args:
        logic: Rule logic dict
        validation_config: Validation configuration

    Returns:
        FilterValidationResult with any errors/warnings
    """
    errors: list[str] = []
    warnings: list[str] = []

    # Extract filter from rule logic
    strategy = logic.get("strategy") if logic else None
    filter_node = strategy.get("filterNode") if strategy else None
    filter_obj = filter_node.get("filter") if filter_node else None

    # Extract all filter types used
    string_types = sorted(extract_string_filter_types(filter_obj))
    numeric_types = sorted(extract_numeric_filter_types(filter_obj))
    null_types = sorted(extract_null_filter_types(filter_obj))

    # Validate string filters
    for filter_type in string_types:
        if filter_type in validation_config.unsupported_string_filters:
            errors.append(
                f'String filter type "{filter_type}" is not supported. '
                f'Use one of: {", ".join(validation_config.supported_string_filters)}'
            )
        elif filter_type not in validation_config.supported_string_filters:
            warnings.append(
                f'String filter type "{filter_type}" is not in the list of tested filters. '
                "It may not render correctly in the UI."
            )

    # Validate numeric filters
    if validation_config.supported_numeric_filters:
        for filter_type in numeric_types:
            if filter_type not in validation_config.supported_numeric_filters:
                warnings.append(
                    f'Numeric filter type "{filter_type}" is not in the list of tested filters.'
                )

    # Validate null filters
    if validation_config.supported_null_filters:
        for filter_type in null_types:
            if filter_type not in validation_config.supported_null_filters:
                warnings.append(
                    f'Null filter type "{filter_type}" is not in the list of tested filters.'
                )

    return FilterValidationResult(
        valid=len(errors) == 0,
        errors=errors,
        warnings=warnings,
        used_filters={
            "string": string_types,
            "numeric": numeric_types,
            "null": null_types,
        },
    )


def get_filter_summary(logic: Any) -> dict[str, Any]:
    """Get a summary of all filter types used."""
    strategy = logic.get("strategy") if logic else None
    filter_node = strategy.get("filterNode") if strategy else None
    filter_obj = filter_node.get("filter") if filter_node else None

    has_compound = False
    if filter_obj and isinstance(filter_obj, dict):
        has_compound = any(
            key in filter_obj for key in ["orFilterRule", "andFilterRule", "notFilterRule"]
        )

    return {
        "string_filters": sorted(extract_string_filter_types(filter_obj)),
        "numeric_filters": sorted(extract_numeric_filter_types(filter_obj)),
        "null_filters": sorted(extract_null_filter_types(filter_obj)),
        "has_compound_filters": has_compound,
    }
